### DISCUSSION TITLE
How are you authorizing users?

### DISCUSSION THREAD

**Discussion by elverskog** [aaronhuggins/pouchdb_deno#10] [General]
Hi. I'm looking into Deno
...and trying to fight back the euphoria of not seeing a baseline of 200 node_modules and @types that may break at any time; amongst a pool of other things I wouldn't miss. That said, I have a lot more digging to do. 

I'm new to couch as well but have been playing around with it browser side.
As Deno is server-side and by looking at the README, it looks like you mainly ported couch to use it server side (without any remote syncing).

I'm just curious what a common use case is.
is the advantage to have a DB inside a node process (speed, access etc) as oppose to say couchdb, however "far away" over HTTP?
Is pouch the system of record in this case?

Thanks for "just doing it yourself" and for any insight you can give.

**Comment by aaronhuggins** [comment]
Hey, thanks for giving this project a look! I have many of the same feelings about Deno. I absolutely love it.

I've been using PouchDB for *years*. It's been great, as I've been able to write an app once, have a local DB working for small deployments and use-cases, and then eventually spin up CouchDB for when I needed remote-only or remote-synced strategies.

This PouchDB port works server-side and with remote CouchDB databases. I'm supporting using the `idb`, `indexeddb`, `memory`, `http`, and `https` adapters. Sync *should* be fully functional, although I have not personally tested it.

My most common use case, personally, is PouchDB is the system of record, period. Any adapter I choose is then the store for those records.

Thinking of PouchDB this way, for me, then allows me to use whatever I want to back my data, whether it's the `memory` adapter for testing and local dev, or `indexeddb` for persist to browser (or disk, in Deno!), or set up a remote CouchDB, or frankly any mix of the above.

This flexibility is what I *love* about PouchDB. It's simple to use, but with some really great superpowers.

**Comment by aaronhuggins** [comment]
Hmm... Not sure I understand the title of this thread.

Might need you to expand on that question.

**Comment by elverskog** [comment]
Hey, thanks for taking the time for a detailed explanation.
A lot to unpack there of course. Still parsing...

Sorry, I realize I didn't really address my question in the body, per auth.
I think what I want to know is how you are granting/restricting access per user to a Deno app and (maybe indirectly) hence to pouch?
For example, are you using pouch's `_users `as a way of creating access controls in a Deno App?
Or are you building/using an auth system purely in Deno that uses one couch user (admin?) globally?

As mentioned I having mainly been trying pouch out client side, so I need to get a cycling token there of course (can't just code in a visible PW). My thinking may be tainted by that.

**Reply by aaronhuggins** [reply]
Ah, I understand. The PouchDB ecosystem has some auth plugins already written targeting Node. I haven't had a use case (yet) to use them with Deno.

If you're interested in getting one of those existing plugins working, feel free to experiment. I would be happy to provide feedback and instruction to get one working.

Here's the plugin I'd recommend, but there are others too.

https://www.npmjs.com/package/pouchdb-auth

**Comment by elverskog** [comment]
Thanks again. Not sure if I will use pouch's users or if I will use some auth system like okta to access the app as a whole.

That seems more of a standard way to run a purely serverside db from what i have done in the past.
Pouch will then just run without a user of course.
Maybe I will have a unique db per user or I'll tag docs somehow, not sure yet.

Again, I was stuck thinking about pouch in a browser/auto-sync-to-couch context.
Which is the "magic" that brought me to pouch in the first place.

_One question, if you can help: Are you using any kind of CLI or UI to view and manipulate your copy of couch.
For example if I built up a series of records that have the wrong format, how could I just drop/clear a whole db?_

Aside from that hurdle, I'm pushing along and have your port working in a build using Deno and Alephjs.
Which is allowing me to build out some FE interactivity,  has HMR and a simple way to build a route based API built in .
Just now building out various CRUD tasks.

**Reply by aaronhuggins** [reply]
> One question, if you can help: Are you using any kind of CLI or UI to view and manipulate your copy of couch.
> For example if I built up a series of records that have the wrong format, how could I just drop/clear a whole db?

For production, when using CouchDB I'll just use their [own interface, Fauxton](https://docs.couchdb.org/en/3.2.2/fauxton/install.html).

If I need to do more autmated stuff, I'll whip up maintenance scripts using PouchDB. I always make sure not to deploy those scripts to my apps, in case there's ever a compromise I don't want someone to have direct access they're not supposed to have.

**Comment by elverskog** [comment]
Agh. sorry I meant to write:
"Are you using any kind of CLI or UI to view and manipulate your copy of **pouch**."

**Reply by aaronhuggins** [reply]
When working locally, I can go into the REPL for Deno and hand type Pouch stuff.

Otherwise, it's back to ad-hoc maintenance scripts.

**Comment by elverskog** [comment]
Thx. Well I can seem to import couch via an eval and run a `new PouchDB` command with no errors (I'm using a local copy this minute so the relative paths were a pain)  but after that I am lost.

If you have a sec to give an example, great, if not no worries (not your job :) )
I can likely create some scripts as needed.
